core/src/test/java/io/grpc/internal/SpiffeUtil2Test.java

```diff
@@ -0,0 +1,151 @@
+import io.grpc.util.CertificateUtils;
+import java.io.IOException;
+import java.nio.file.NoSuchFileException;
+import java.security.cert.CertificateException;
+import java.security.cert.CertificateParsingException;
+import java.security.cert.X509Certificate;
+import java.util.List;
+import org.junit.Before;
+import org.junit.Test;
+
+public class SpiffeUtil2Test {
+
+  private static final String SPIFFE_PEM_FILE = "spiffe_cert.pem";
+  private static final String SPIFFE_MULTI_VALUES_PEM_FILE = "spiffe_cert_multi.pem";
+  private static final String SERVER_0_PEM_FILE = "server0.pem";
+  private static final String TEST_DIRECTORY_PREFIX = "io/grpc/internal/";
+  private static final String SPIFFE_TRUST_BUNDLE_FILE = "spiffebundle.json";
+  private static final String SPIFFE_TRUST_BUNDLE_MALFORMED = "spiffebundle_malformed.json";
+  private static final String SPIFFE_TRUST_BUNDLE_WRONG_ELEMENTS =
+      "spiffebundle_wrong_elements.json";
+  private static final String SPIFFE_TRUST_BUNDLE_DUPLICATES = "spiffebundle_duplicates.json";
+  private static final String SPIFFE_TRUST_BUNDLE_WITH_WRONG_ROOT =
+      "spiffebundle_wrong_root.json";
+
+
+  private X509Certificate[] spiffeCert;
+  private X509Certificate[] spiffeMultiCert;
+  private X509Certificate[] serverCert0;
+
+
+  @Before
+  public void setUp() throws CertificateException {
+    spiffeCert = CertificateUtils.getX509Certificates(TlsTesting.loadCert(SPIFFE_PEM_FILE));
+    spiffeMultiCert = CertificateUtils.getX509Certificates(TlsTesting
+        .loadCert(SPIFFE_MULTI_VALUES_PEM_FILE));
+    serverCert0 = CertificateUtils.getX509Certificates(TlsTesting.loadCert(SERVER_0_PEM_FILE));
+  }
+
+  @Test
+  public void extractSpiffeIdSuccessTest() throws CertificateParsingException {
+    Optional<SpiffeUtil2.SpiffeId> spiffeId = SpiffeUtil2.extractSpiffeId(spiffeCert);
+    assertEquals("foo.bar.com", spiffeId.get().getTrustDomain());
+    assertEquals("client/workload/1", spiffeId.get().getPath());
+  }
+
+  @Test
+  public void extractSpiffeIdFailureTest() throws CertificateParsingException {
+    Optional<SpiffeUtil2.SpiffeId> spiffeId = SpiffeUtil2.extractSpiffeId(serverCert0);
+    assertFalse(spiffeId.isPresent());
+    IllegalArgumentException iae = assertThrows(IllegalArgumentException.class, () -> SpiffeUtil2
+        .extractSpiffeId(spiffeMultiCert));
+    assertEquals("Multiple URI SAN values found in the leaf cert.", iae.getMessage());
+
+  }
+
+  @Test
+  public void extractSpiffeIdFromChainTest() throws CertificateParsingException {
+    X509Certificate[] leafWithSpiffeChain = new X509Certificate[]{spiffeCert[0], serverCert0[0]};
+    assertTrue(SpiffeUtil2.extractSpiffeId(leafWithSpiffeChain).isPresent());
+    X509Certificate[] leafWithoutSpiffeChain = new X509Certificate[]{serverCert0[0], spiffeCert[0]};
+    assertFalse(SpiffeUtil2.extractSpiffeId(leafWithoutSpiffeChain).isPresent());
+  }
+
+  @Test
+  public void extractSpiffeIdParameterValidityTest() {
```

Comment:
Can we add a test where we fail to extract a SPIFFE ID from a cert without any URI SANs?

Reply:
We don't 'hard fail' in that case, but just return `Optional.absent` - https://github.com/grpc/grpc-java/pull/11575/files#diff-fdbe5016403200b9fc7adab29732582366f2494986933098d86fddd1081d6a0bR258